pkg/util/chunk/chunk_in_disk.go

```diff
@@ -0,0 +1,320 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"strconv"
+	"unsafe"
+
+	errors2 "github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/parser/terror"
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util/disk"
+	"github.com/pingcap/tidb/pkg/util/memory"
+)
+
+const byteLen = int64(unsafe.Sizeof(byte(0)))
+const intLen = int64(unsafe.Sizeof(int(0)))
+const int64Len = int64(unsafe.Sizeof(int64(0)))
+
+const chkFixedSize = intLen + 3*intLen
+const colMetaSize = int64Len * 4
+
+// DataInDiskByChunks represents some data stored in temporary disk.
+// They can only be restored by chunks.
+type DataInDiskByChunks struct {
+	fieldTypes        []*types.FieldType
+	offsetOfEachChunk []int64
+
+	totalDataSize int64
+	totalRowNum   int64
+	diskTracker   *disk.Tracker // track disk usage.
+
+	dataFile diskFileReaderWriter
+
+	// Write or read data needs this buffer to temporarily store data
+	buf []byte
+}
+
+// NewDataInDiskByChunks creates a new DataInDiskByChunks with field types.
+func NewDataInDiskByChunks(fieldTypes []*types.FieldType) *DataInDiskByChunks {
+	l := &DataInDiskByChunks{
+		fieldTypes:    fieldTypes,
+		totalDataSize: 0,
+		totalRowNum:   0,
+		// TODO: set the quota of disk usage.
+		diskTracker: disk.NewTracker(memory.LabelForChunkDataInDiskByChunks, -1),
+		buf:         make([]byte, 0, 4096),
+	}
+	return l
+}
+
+func (d *DataInDiskByChunks) initDiskFile() (err error) {
+	err = disk.CheckAndInitTempDir()
+	if err != nil {
+		return
+	}
+	err = d.dataFile.initWithFileName(defaultChunkDataInDiskByRowsPath + strconv.Itoa(d.diskTracker.Label()))
+	if err != nil {
+		return
+	}
+	return
+}
+
+// GetDiskTracker returns the memory tracker of this List.
+func (d *DataInDiskByChunks) GetDiskTracker() *disk.Tracker {
+	return d.diskTracker
+}
+
+// Add adds a chunk to the DataInDiskByChunks. Caller must make sure the input chk has the same field types.
+// Warning: Do not concurrently call this function.
+func (d *DataInDiskByChunks) Add(chk *Chunk) (err error) {
+	if chk.NumRows() == 0 {
+		return errors2.New("Chunk spilled to disk should have at least 1 row")
+	}
+
+	if d.dataFile.file == nil {
+		err = d.initDiskFile()
+		if err != nil {
+			return
+		}
+	}
+
+	serializedBytesNum := d.serializeDataToBuf(chk)
+
+	var writeNum int
+	writeNum, err = d.dataFile.write(d.buf)
```

Comment:
Is it necessary to write twice here? First write the chunk to buf, and then write from buf to file. Why can't it be written directly to file?

Reply:
> Is it necessary to write twice here? First write the chunk to buf, and then write from buf to file. Why can't it be written directly to file?

I think it's necessary or we have to spill a chunk to disk with io for several times.